dask/array/core.py

```diff
@@ -893,20 +897,72 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
         except AttributeError:
             dsk = {}
 
-        update = insert_to_ooc(src, tgt, lock=lock, region=reg)
-        keys.extend(update)
+            src = Array(
+                src.__dask_optimize__(
+                    src.__dask_graph__(),
+                    src.__dask_keys__()
+                ),
+                src.name,
+                src.chunks,
+                src.dtype,
+                src.shape
+            )
```

Comment:
The previous `src` object here is a dask array?  If so then is there a reason why we call optimize before passing it to another array?  We should only need to call optimize when we leave the array collection type.

Reply:
Sorry not following. This was my understanding of your suggestion in the thread above.

There is actually only one case where we generate a Dask Array and I'm not sure that we shouldn't optimize it in that case too given we are trying to write this efficiently.

Reply:
The danger is when we take a graph out of a dask.array and put it into a delayed object, or just compute it.  Then the array-specific optimizations don't run.  

There is relatively little danger when putting a task graph into a dask array, unless that task graph comes from some other collection with non-trivial optimizations (like bag or dataframe)

Reply:
Right. So the only case that I can think of that wouldn't need this optimization would be `compute=False, keep=True`. Would you like a guard against optimizing in that case? All the other cases create `Delayed` object(s).

Reply:
When we create those delayed objects we point them to graphs that have array computations in them.  Have those array tasks already gone through an optimization phase?

### This is bad

```python
x = da.from_array(...)
x[:10][:5]  # this will be optimized away

v = Delayed(key, x.dask)  # lose array optimizations at this point
```

### This is good

```python
x = da.from_array(...)
x[:10][:5]  # this will be optimized away

dsk = x.__dask_optimize__(x.__dask_graph__(), x.__dask_keys__())  # pre-optimize graph

v = Delayed(key, dsk) 
```

Reply:
Right. That's what this optimize call is doing.

Reply:
OK, I see, you're optimizing and then placing it back into the dask array.

What happens if the except branch is not triggered?  

Hrm, one issue that will come up when storing multiple arrays is that if there are multiple sources that have overlapping computations then optimizing them separately will likely create redundant work (dask arrays can fuse operations somewhat aggressively).  It looks like optimizing everything ahead of time might be a bit painful.

Reply:
Yeah, we still need a Dask Array object in `insert_to_ooc`. Though as we are just pulling out `keys` and finding corresponding `slice`s, I guess we could forego the Dask Array if we changed `insert_to_ooc`'s signature. Is there any performance penalty when stuffing this back into a Dask Array?

Are you referring to `except AttributeError` above? I think that is from a PR that allowed `Delayed` objects as targets. So my understanding is it was never meant to be anything more sophisticated than say a function that created a place to store the results. Honestly not sure it would even work with something more sophisticated than that even in its current form in `master`.

I see. So would it be best to get Dask graphs for everything before the `for`-loop and optimize them? Let me give that some thought. I think it should be possible.

Reply:
>  Is there any performance penalty when stuffing this back into a Dask Array?

I *think* that this should be more-or-less fine.  We should profile this on a many-chunked array before merging though and see if any new hot-spots arise.

Reply:
> What happens if the except branch is not triggered?

Had an indentation error that I overlooked. 😞 Have since fixed it.

Reply:
> Hrm, one issue that will come up when storing multiple arrays is that if there are multiple sources that have overlapping computations then optimizing them separately will likely create redundant work (dask arrays can fuse operations somewhat aggressively). It looks like optimizing everything ahead of time might be a bit painful.

Tweaked the optimization to handle this case. Honestly not to bad since `__dask_optimize__` is a static method. Also the docs point out this can work on multiple collections. So this was basically designed to work. :)